### PR TITLE
added a duplicate version name check for models

### PR DESCRIPTION
**Pull request by ARADDCC002**

### CODE REVIEW

**Comment by ARADDCC002** [comment]
test

File: server/routes/v1/upload.ts [hunk 1]
```diff
@@ -107,7 +107,7 @@ export const postUpload = [
           metadata: metadata,
         })
       } catch (err: any) {
-        if (err.toString().includes('duplicate key error')) {
+        if (err.toString().includes('Duplicate version name for model')) {
```
Comment (a3957273) [inline server/routes/v1/upload.ts:110]: We should be testing the error status code, instead of the text.

File: server/services/version.ts [hunk 2]
```diff
@@ -93,6 +93,14 @@ export async function createVersion(user: UserDoc, data: CreateVersion) {
     throw Forbidden({ data }, 'Unable to create version, failed permissions check.')
   }
 
+  let allVersionsForModel = VersionModel.find({ model: version.model })
```
Comment (a3957273) [inline server/services/version.ts:96]: Should be const.

File: server/services/version.ts [hunk 3]
```diff
@@ -93,6 +93,14 @@ export async function createVersion(user: UserDoc, data: CreateVersion) {
...
+  if (
+    (await allVersionsForModel).filter((versionDuplicateCheck) => versionDuplicateCheck.version === version.version)
```
Comment (a3957273) [inline server/services/version.ts:98]: You shouldn't be doing this logic on the server, you should be getting Mongo to only find models with the specific model _and_ the specified version.

**Review by a3957273 — CHANGES_REQUESTED** [review]

File: server/services/version.ts (hunk 2 above)
Comment (ARADDCC002) [reply server/services/version.ts:96]: done

File: server/routes/v1/upload.ts (hunk 1 above)
Comment (ARADDCC002) [reply server/routes/v1/upload.ts]: Resolved

File: server/services/version.ts (hunk 3 above)
Comment (ARADDCC002) [reply server/services/version.ts]: Resolved

File: server/routes/v1/upload.ts [hunk 4]
```diff
@@ -40,7 +40,8 @@ export const postUpload = [
     const session = await mongoose.startSession()
     return session.withTransaction(async () => {
       const files = req.files as unknown as MinioFile
-      const mode = req.query.mode
+      const mode: string = req.query.mode as string
```
Comment (a3957273) [inline server/routes/v1/upload.ts:43]: I don't think you need `: string` here, as TypeScript should already know it's a string.

File: server/routes/v1/upload.ts [hunk 5]
```diff
@@ -102,12 +103,16 @@ export const postUpload = [
...
+          req.user!,
+          {
+            version: metadata.highLevelDetails.modelCardVersion,
+            metadata: metadata,
+          },
+          modelUuid
+        )
       } catch (err: any) {
-        if (err.toString().includes('duplicate key error')) {
+        if (err.code === 400) {
```
Comment (a3957273) [inline server/routes/v1/upload.ts:115]: Sorry, clearly didn't mean code.  I was vaguely expecting the code `E11000` we use elsewhere (and is documented https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml).  Which is the unique error code so we only catch the duplicate error here.  It should be somewhere on the response error we get?

Also, if it isn't this error we want to rethrow the error, e.g.:

```
} catch(err: any) {
  if (err.some.code === 'E11000') {
    return res.status(409.json({ ... })
  }

  throw err
}
```

File: server/services/version.ts [hunk 6]
```diff
@@ -86,13 +87,22 @@ interface CreateVersion {
   metadata: any
 }
 
-export async function createVersion(user: UserDoc, data: CreateVersion) {
+export async function createVersion(user: UserDoc, data: CreateVersion, parentUuid: string) {
```
Comment (a3957273) [inline server/services/version.ts:90]: Services shouldn't interact with other domains.  E.g. here, the `Version` service interacts with the `Model` domain.  Instead, you should use the `Model` service (or expect the calling process to pass in a full Model instance).

**Review by a3957273 — CHANGES_REQUESTED** [review]

File: server/routes/v1/upload.ts (hunk 4 above)
Comment (ARADDCC002) [reply server/routes/v1/upload.ts:43]: Resvoled

**Review by a3957273 — APPROVED** [review]

**Pull request merged**